src/Databases/DataLake/DatabaseDataLake.cpp

```diff
@@ -448,20 +448,62 @@ DatabaseTablesIteratorPtr DatabaseDataLake::getTablesIterator(
     const FilterByNameFunction & filter_by_table_name,
     bool /* skip_not_loaded */) const
 {
+    std::mutex mutex;
     Tables tables;
     auto catalog = getCatalog();
     const auto iceberg_tables = catalog->getTables();
 
+    auto & pool = Context::getGlobalContextInstance()->getIcebergCatalogThreadpool();
+
+    std::vector<std::shared_ptr<std::promise<void>>> promises;
+    std::vector<std::future<void>> futures;
     for (const auto & table_name : iceberg_tables)
     {
         if (filter_by_table_name && !filter_by_table_name(table_name))
             continue;
+        try
+        {
+            promises.emplace_back(std::make_shared<std::promise<void>>());
+            futures.emplace_back(promises.back()->get_future());
+
+            pool.scheduleOrThrow(
+                [this, table_name, &tables, &mutex, context_, promise=promises.back()]() mutable
+                {
+                    try
+                    {
+                        auto storage = tryGetTable(table_name, context_);
+                        std::lock_guard lock(mutex);
+                        [[maybe_unused]] bool inserted = tables.emplace(table_name, storage).second;
+                        chassert(inserted);
+                        promise->set_value();
+                    }
+                    catch (...)
+                    {
+                        tryLogCurrentException(log, fmt::format("Ignoring table {}", table_name));
+                        promise->set_exception(std::current_exception());
+                    }
+                });
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Failed to schedule task");
+            pool.wait();
+
+            /// Rethrow exceptions if any
+            for (auto & future : futures)
```

Comment:
If future was added to vector but task was not scheduled in the pool then it is a deadlock